File: src/main/java/io/femo/http/drivers/DefaultHttpsServer.java
package io.femo.http.drivers;

import io.femo.http.HttpException;
import io.femo.http.HttpServer;
import io.femo.http.HttpsServer;
import io.femo.http.drivers.server.HttpsServerThread;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.net.ssl.KeyManagerFactory;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManagerFactory;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.security.*;
import java.security.cert.CertificateException;

/**
 * Created by felix on 9/14/16.
 */
public class DefaultHttpsServer extends DefaultHttpServer implements HttpsServer {

    private static final Logger LOGGER = LoggerFactory.getLogger("HTTP");

    private String keystore;
    private String keystorePass;
    private String keyPass;

    public DefaultHttpsServer(int port) {
        super(port, true);
    }

    public DefaultHttpsServer(DefaultHttpServer server) {
        super(server.port, true);
        this.httpHandlerStack = server.httpHandlerStack;
    }

    @Override
    public HttpsServer keystorePass(String keystorePass) {
        this.keystorePass = keystorePass;
        return this;
    }

    @Override
    public HttpsServer keyPass(String keyPass) {
        this.keyPass = keyPass;
        return this;
    }

    @Override
    public HttpsServer keystore(String keystore) {
        this.keystore = keystore;
        return this;
    }

    @Override
    public HttpsServer start() {
        SSLContext sslContext;
        try {
            sslContext = SSLContext.getInstance("TLSv1.2");
            KeyStore keyStore = KeyStore.getInstance("JKS");
            FileInputStream fileInputStream = new FileInputStream(keystore);
            keyStore.load(fileInputStream, keystorePass.toCharArray());

            KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance("SunX509");
            keyManagerFactory.init(keyStore, keyPass.toCharArray());

            TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance("SunX509");
            trustManagerFactory.init(keyStore);

            sslContext.init(keyManagerFactory.getKeyManagers(), trustManagerFactory.getTrustManagers(), null);
        } catch (NoSuchAlgorithmException | KeyStoreException | CertificateException |
                IOException | UnrecoverableKeyException | KeyManagementException e) {
            LOGGER.error("Could not initialize security context", e);
            return this;
        }
        try {
            this.serverThread = new HttpsServerThread(httpHandlerStack, sslContext);
        } catch (NoSuchAlgorithmException | KeyManagementException e) {
            LOGGER.error("Could not start secure context", e);
            return this;
        }
        super.start();
        return this;
    }
}

File: src/main/java/org/web25/http/drivers/DefaultHttpsServer.kt
package org.web25.http.drivers

import org.slf4j.LoggerFactory
import org.web25.http.HttpContext
import org.web25.http.drivers.server.HttpsServerThread
import org.web25.http.server.Configurator
import org.web25.http.server.HttpsServer
import java.io.FileInputStream
import java.io.IOException
import java.security.*
import java.security.cert.CertificateException
import javax.net.ssl.KeyManagerFactory
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManagerFactory

/**
 * Created by felix on 9/14/16.
 */
class DefaultHttpsServer : DefaultHttpServer, HttpsServer {

    private var keystore: String? = null
    private var keystorePass: String? = null
    private var keyPass: String? = null

    constructor(port: Int, context: HttpContext) : super(port, true, context) {}

    constructor(server: DefaultHttpServer, context: HttpContext) : super(server.port, true, context) {
        this.httpHandlerStack = server.httpHandlerStack
    }

    constructor(configurator: Configurator, context: HttpContext) : this(configurator.getInt("port"), context)

    override fun keystorePass(keystorePass: String): HttpsServer {
        this.keystorePass = keystorePass
        return this
    }

    override fun keyPass(keyPass: String): HttpsServer {
        this.keyPass = keyPass
        return this
    }

    override fun keystore(keystore: String): HttpsServer {
        this.keystore = keystore
        return this
    }

    override fun start(): HttpsServer {
        val sslContext: SSLContext
        try {
            sslContext = SSLContext.getInstance("TLSv1.2")
            val keyStore = KeyStore.getInstance("JKS")
            val fileInputStream = FileInputStream(keystore!!)
            keyStore.load(fileInputStream, keystorePass!!.toCharArray())

            val keyManagerFactory = KeyManagerFactory.getInstance("SunX509")
            keyManagerFactory.init(keyStore, keyPass!!.toCharArray())

            val trustManagerFactory = TrustManagerFactory.getInstance("SunX509")
            trustManagerFactory.init(keyStore)

            sslContext.init(keyManagerFactory.keyManagers, trustManagerFactory.trustManagers, null)
        } catch (e: NoSuchAlgorithmException) {
            LOGGER.error("Could not initialize security context", e)
            return this
        } catch (e: KeyStoreException) {
            LOGGER.error("Could not initialize security context", e)
            return this
        } catch (e: CertificateException) {
            LOGGER.error("Could not initialize security context", e)
            return this
        } catch (e: IOException) {
            LOGGER.error("Could not initialize security context", e)
            return this
        } catch (e: UnrecoverableKeyException) {
            LOGGER.error("Could not initialize security context", e)
            return this
        } catch (e: KeyManagementException) {
            LOGGER.error("Could not initialize security context", e)
            return this
        }

        try {
            this.serverThread = HttpsServerThread(httpHandlerStack, sslContext, context)
        } catch (e: NoSuchAlgorithmException) {
            LOGGER.error("Could not start secure context", e)
            return this
        } catch (e: KeyManagementException) {
            LOGGER.error("Could not start secure context", e)
            return this
        }

        super.start()
        return this
    }

    companion object {

        private val LOGGER = LoggerFactory.getLogger("HTTP")
    }
}
